app/models/enterprise.rb

```diff
@@ -470,6 +470,8 @@ def restore_permalink
   end
 
   def initialize_permalink
+    return unless name
```

Comment:
I'd rather make the name column no-nullable at DB level to mimic the model validation and so we are 100% sure of its presence and avoid the defensive coding. Was this necessary for this PR to work?

Reply:
> Was this necessary for this PR to work?

@sauloperez Yeah. Without the guard clause, the `it { is_expected.to validate_presence_of(:name) }` spec fails.

```
1) Enterprise validations is expected to validate that :name cannot be empty/falsy
     Failure/Error: test_permalink = test_permalink.parameterize
     
     NoMethodError:
       undefined method `parameterize' for nil:NilClass
     # ./app/models/enterprise.rb:355:in `find_available_permalink'
     # ./app/models/enterprise.rb:473:in `initialize_permalink'
     # ./spec/models/enterprise_spec.rb:114:in `block (3 levels) in <top (required)>'
```

Reply:
hey, I think this is incorrect.

If you look at enterprise.rb above you see:
```
  def self.find_available_permalink(test_permalink)
    test_permalink = test_permalink.parameterize
    test_permalink = "my-enterprise" if test_permalink.blank?

```
basically, if no name is available, "my-enterprise will be used. I dont think we should simply remove that path because if the test. I think we need to adapt the test.

Reply:
> basically, if no name is available, "my-enterprise" will be used.

That's the case if `test_permalink` is an empty string. If it's `nil`, the code will fail with the `undefined method 'parameterize' for nil:NilClass` error. Why should we bother setting a permalink if the record is invalid (without a name)?

Reply:
I'd revert the question: why do we need to change the logic just to fix the tests?
I think, I am not sure, there may be a case where a product with no name will need a permalink. For example, the permalink is the string used on the url to edit a product.

Reply:
> why do we need to change the logic just to fix the tests?

@luisramos0 We're currently [testing the presence validation on an existing enterprise record](https://github.com/openfoodfoundation/openfoodnetwork/blob/ebd7e9599b516380da26d2885139cb0dfa8d4c2c/spec/models/enterprise_spec.rb#L112-L113) and so the permalink is set before validations in the `initialize_permalink` method since `permalink` is `nil` — note the lambda passed to the `before_validation` call [here](https://github.com/openfoodfoundation/openfoodnetwork/blob/ebd7e9599b516380da26d2885139cb0dfa8d4c2c/app/models/enterprise.rb#L101). When shoulda matchers tests the validation by setting the name to `nil`, the `initialize_permalink` method isn't called again because the permalink is already set (the lambda returns `false`). That's why we weren't able to expose the `undefined method 'parameterize' for nil:NilClass` error previously.

To reproduce the issue, try running the following code in the rails console (on the master branch):

```
enterprise = Enterprise.new
enterprise.valid?
```
We don't have to create a record to test the presence validation (you can check [the ActiveModel validation matchers example on the Shoulda matchers website](http://matchers.shoulda.io/#tab1) if you want to). When I removed [the enterprise creation](https://github.com/openfoodfoundation/openfoodnetwork/blob/ebd7e9599b516380da26d2885139cb0dfa8d4c2c/spec/models/enterprise_spec.rb#L112), I noticed the test started failing because the `initialize_permalink` method was called on an instance with a name set to `nil`. To fix it, a guard clause was added.

> I think, I am not sure, there may be a case where a product with no name will need a permalink. For example, the permalink is the string used on the url to edit a product.

Not sure if I understand. We're in the context of enterprises and If an enterprise has no name, it is invalid and we don't really have to set a permalink for an invalid record.

https://github.com/openfoodfoundation/openfoodnetwork/blob/ebd7e9599b516380da26d2885139cb0dfa8d4c2c/app/models/enterprise.rb#L92

Do let me know if I'm missing something.

Reply:
ok, I see what problem you are trying to solve with the guard clause.

imho if for some reason the enterprise has no name we can still use the default permalink calculation logic in find_available_permalink. Would this not be better approach?
![image](https://user-images.githubusercontent.com/1640378/97198323-9367de80-17a6-11eb-90fe-f82bf228fad7.png)

Anyway, this is a detail we shouldnt spend too much time around. Also, I now dont think the guard clause will cause any trouble in prod, so I am approving this PR as is :+1:

Reply:
> Would this not be better approach?

We could do that but it would send an unnecessary database request to set an attribute for an invalid enterprise which doesn't make sense.

> Anyway, this is a detail we shouldnt spend too much time around. Also, I now dont think the guard clause will cause any trouble in prod, so I am approving this PR as is 👍

Thanks! 🚀